Index: src/compiler/scala/tools/nsc/Global.scala
===================================================================
--- src/compiler/scala/tools/nsc/Global.scala	(revision 18542)
+++ src/compiler/scala/tools/nsc/Global.scala	(working copy)
@@ -49,6 +49,8 @@
   //def this() = this(new Settings, new ConsoleReporter)
 
   // sub-components --------------------------------------------------
+  var selectTypeErrors :Map[Tree, Type] = Map()
+  def resetSelectTypeErrors = selectTypeErrors = Map()
 
   /** Print tree in detailed form */
   object nodePrinters extends {
Index: src/compiler/scala/tools/nsc/typechecker/Typers.scala
===================================================================
--- src/compiler/scala/tools/nsc/typechecker/Typers.scala	(revision 18542)
+++ src/compiler/scala/tools/nsc/typechecker/Typers.scala	(working copy)
@@ -3203,6 +3203,7 @@
         if (!sym.exists) {
           if (settings.debug.value) Console.err.println("qual = "+qual+":"+qual.tpe+"\nSymbol="+qual.tpe.termSymbol+"\nsymbol-info = "+qual.tpe.termSymbol.info+"\nscope-id = "+qual.tpe.termSymbol.info.decls.hashCode()+"\nmembers = "+qual.tpe.members+"\nname = "+name+"\nfound = "+sym+"\nowner = "+context.enclClass.owner)
           if (!qual.tpe.widen.isErroneous) {
+            global.selectTypeErrors += (tree -> qual.tpe.widen)
             error(tree.pos,
               if (name == nme.CONSTRUCTOR) 
                 qual.tpe.widen+" does not have a constructor"
